ui/admin/app/routes/scopes/scope/roles/role/scopes.js

```diff
@@ -40,49 +41,61 @@ export default class ScopesScopeRolesRoleScopesRoute extends Route {
    * Loads grant scopes for current role.
    * @return {Promise<{role: RoleModel, grantScopes: [ScopeModel], allGrantScopes: [ScopeModel], totalItems: number}> }
    */
-  async model({ search, parentScopes, types, page, pageSize }) {
-    const role = this.modelFor('scopes.scope.roles.role');
-    const filters = {
-      scope_id: [],
-      type: [],
-      id: [],
-    };
-    role.grant_scope_ids.forEach((id) => {
-      if (id.startsWith('p_') || id.startsWith('o_')) {
-        filters.id.push({ equals: id });
+  async model(params) {
+    const useDebounce =
+      this.retrieveData?.lastPerformed?.args?.[0].search !== params.search;
+    return this.retrieveData.perform({ ...params, useDebounce });
+  }
+
+  retrieveData = restartableTask(
+    async ({ search, parentScopes, types, page, pageSize, useDebounce }) => {
+      if (useDebounce) {
+        await timeout(250);
       }
-    });
 
-    let grantScopes = [];
-    if (!parentScopes.length && !types.length) {
-      role.grantScopeKeywords.forEach((keyword) => {
-        if (keyword.includes(search)) {
-          grantScopes.push({ id: keyword });
+      const role = this.modelFor('scopes.scope.roles.role');
+      const filters = {
+        scope_id: [],
+        type: [],
+        id: [],
+      };
+      role.grant_scope_ids.forEach((id) => {
+        if (id.startsWith('p_') || id.startsWith('o_')) {
+          filters.id.push({ equals: id });
         }
       });
-    }
 
-    let allGrantScopes = [];
-    let totalItems = grantScopes.length;
-    if (filters.id.length) {
-      parentScopes.forEach((id) => filters.scope_id.push({ equals: id }));
-      types.forEach((type) => filters.type.push({ equals: type }));
+      let grantScopes = [];
+      if (!parentScopes.length && !types.length) {
+        role.grantScopeKeywords.forEach((keyword) => {
+          if (keyword.includes(search) || !search) {
```

Comment:
There was a slight change in logic here, I added the `|| !search` because I was surprised to see that changing the initial state of `search` from empty string to undefined broke tests.

It turns out we were depending on `keyword.includes(search)` when search was `''` to include all the keywords by default which seemed unexpected so I made this more explicit.